CoreScriptsRoot/Modules/NewChat.lua

```diff
@@ -0,0 +1,233 @@
+		{
+			Visible = true,
+			MessageCount = 0,
+			TopbarEnabled = true,
+			CoreGuiEnabled = true,
+		}
+
+		local communicationsConnections = {}
+		local eventConnections = {}
+
+		local MakeSystemMessageCache = {}
+
+		local function FindIndexInCollectionWithType(collection, indexName, type)
+			if (collection and collection[indexName] and collection[indexName]:IsA(type)) then
+				return collection[indexName]
+			end
+			return nil
+		end
+
+		function moduleApiTable:ToggleVisibility()
+			ChatWindowState.Visible = not ChatWindowState.Visible
+			local event = FindIndexInCollectionWithType(communicationsConnections.ChatWindow, "ToggleVisibility", "BindableEvent")
+			if (event) then
+				event:Fire()
+			else
+				moduleApiTable.VisibilityStateChanged:fire(ChatWindowState.Visible)
+			end
+		end
+
+		function moduleApiTable:SetVisible(visible)
+			ChatWindowState.Visible = visible
+			local event = FindIndexInCollectionWithType(communicationsConnections.ChatWindow, "SetVisible", "BindableEvent")
+			if (event) then
+				event:Fire(ChatWindowState.Visible)
+			else
+				moduleApiTable.VisibilityStateChanged:fire(ChatWindowState.Visible)
+			end
+		end
+
+		function moduleApiTable:FocusChatBar()
+			local event = FindIndexInCollectionWithType(communicationsConnections.ChatWindow, "FocusChatBar", "BindableEvent")
+			if (event) then
+				event:Fire()
+			end
+		end
+
+		function moduleApiTable:GetVisibility()
+			local func = FindIndexInCollectionWithType(communicationsConnections.ChatWindow, "GetVisibility", "BindableFunction")
+			if (func) then
+				return func:Invoke()
+			end
+			return ChatWindowState.Visible
+		end
+
+		function moduleApiTable:GetMessageCount()
+			local func = FindIndexInCollectionWithType(communicationsConnections.ChatWindow, "GetMessageCount", "BindableFunction")
+			if (func) then
+				return func:Invoke()
+			end
+			return ChatWindowState.MessageCount
+		end
+
+		function moduleApiTable:TopbarEnabledChanged(enabled)
+			ChatWindowState.TopbarEnabled = enabled
+			local event = FindIndexInCollectionWithType(communicationsConnections.ChatWindow, "TopbarEnabledChanged", "BindableEvent")
+			if (event) then
+				event:Fire(ChatWindowState.TopbarEnabled)
+			end
+		end
+
+		function moduleApiTable:IsFocused(useWasFocused)
+			local func = FindIndexInCollectionWithType(communicationsConnections.ChatWindow, "IsFocused", "BindableFunction")
+			if (func) then
+				return func:Invoke(useWasFocused)
+			end
+			return false
+		end
+		
+		moduleApiTable.ChatBarFocusChanged = Util.Signal()
+		moduleApiTable.VisibilityStateChanged = Util.Signal()
+		moduleApiTable.MessagesChanged = Util.Signal()
+
+		StarterGui.CoreGuiChangedSignal:connect(function(coreGuiType, enabled)
+			if (coreGuiType == Enum.CoreGuiType.All or coreGuiType == Enum.CoreGuiType.Chat) then
+				ChatWindowState.CoreGuiEnabled = enabled
+
+				local event = FindIndexInCollectionWithType(communicationsConnections.ChatWindow, "CoreGuiEnabled", "BindableEvent")
```

Comment:
This event dispatching code seems to be very similar, we should refactor it into its own function